File: xxf_arch/src/main/java/com/xxf/arch/exception/PermissionDeniedException.java
package com.xxf.arch.exception;

/**
 * @Description: 权限拒绝异常
 * @Author: XGod  xuanyouwu@163.com  17611639080  https://github.com/NBXXF     https://blog.csdn.net/axuanqq
 * @CreateDate: 2020/11/30 19:05
 */
public class PermissionDeniedException extends RuntimeException {
    final String[] permission;

    public String[] getPermission() {
        return permission;
    }

    public PermissionDeniedException(String... permission) {
        super(convertDesc(permission));
        this.permission = permission;
    }

    public static final String convertDesc(String... permissions) {
        StringBuilder permissionSb = new StringBuilder("");
        if (permissions != null && permissions.length > 0) {
            for (int i = 0; i < permissions.length; i++) {
                /**
                 * android.permission.RECORD_AUDIO
                 */
                String per = permissions[i];
                if (i != 0) {
                    permissionSb.append(",");
                }
                permissionSb.append(per.replaceAll("android.permission.", ""));
            }
        }
        return String.format("%s Permission Denied", permissionSb.toString());
    }
}

File: core/lib_permission/src/main/java/com/xxf/permission/PermissionDeniedException.kt
package com.xxf.permission

import android.content.Context
import java.util.*

/**
 * @Description: 权限拒绝异常
 * @Author: XGod  xuanyouwu@163.com  17611639080  https://github.com/NBXXF     https://blog.csdn.net/axuanqq
 * @CreateDate: 2020/11/30 19:05
 */
class PermissionDeniedException(val context: Context, vararg val permission: String) :
    RuntimeException(convertDesc(context, *permission)) {

    /**
     * 获取权限结果
     *
     * @return
     */
    fun getPermissionResult(context: Context): Map<String, Boolean> {
        return getPermissionResult(context, *permission)
    }

    companion object {
        fun getPermissionResult(
            context: Context,
            vararg permissionStr: String
        ): Map<String, Boolean> {
            return context.checkSelfPermissionForResult(*permissionStr)
        }

        /**
         * 将权限被拒绝的组装成描述
         *
         * @param permissionArr
         * @return
         */
        fun convertDesc(context: Context, vararg permissionArr: String): String {
            val permissionResult = getPermissionResult(context, *permissionArr)
            val joinToString = permissionResult.filter {
                !it.value
            }.keys.joinToString(",") {
                it.replace("android.permission.", "")
            }
            return String.format("%s Permission Denied", joinToString)
        }
    }
}